src/contractor.jl

```diff
@@ -40,11 +47,50 @@ function Base.show(io::IO, C::Contractor{N,Nout,F1,F2,ex}) where {N,Nout,F1,F2,e
 end
 
     (C::Contractor)(X) = C.forward(X)[1]
+    (C::BasicContractor)(X) = C.forward(X)[1]
+
+function (C::Contractor)(
+    A::IntervalBox{Nout,T}, X::IntervalBox{N,T})where {N,Nout,T}
+
+    output, intermediate = Forward(C)(X)
+
+    # @show output
+    # @show intermediate
+
+    output_box = IntervalBox(output)
+    constrained = output_box ∩ A
+
+    # if constrained is already empty, eliminate call to backward propagation:
+
+    if isempty(constrained)
+        return emptyinterval(X)
+    end
+
+    # @show X
+    # @show constrained
+    # @show intermediate
+    # @show C.backward(X, constrained, intermediate)
+    return IntervalBox{N,T}(Backward(C)(X, constrained, intermediate) )
+
+end
+
+# allow 1D contractors to take Interval instead of IntervalBox for simplicty:
+
+(C::Contractor)(A::Interval{T}, X::IntervalBox{N,T}) where {N,T} = C(IntervalBox(A), X)
+
+
+function Forward(C::AbstractContractor)
+        return C.forward
+end
+
+function Backward(C::AbstractContractor)
+        return C.backward
+end
 
-function (C::Contractor{N,Nout,F1,F2,ex})(
-    A::IntervalBox{Nout,T}, X::IntervalBox{N,T}) where {N,Nout,F1,F2,ex,T}
+function (C::BasicContractor)(
```

Comment:
This is a duplication of the code at line 52. This is what needs to be factored out.